pkg/reconciler/taskrun/metrics.go

```diff
@@ -0,0 +1,295 @@
+
+	runningTRsCount = stats.Float64("running_taskruns_count",
+		"Number of taskruns executing currently",
+		stats.UnitDimensionless)
+
+	podLatency = stats.Float64("taskruns_pod_latency",
+		"scheduling latency for the taskruns pods",
+		stats.UnitMilliseconds)
+)
+
+type Recorder struct {
+	initialized bool
+
+	task        tag.Key
+	taskRun     tag.Key
+	namespace   tag.Key
+	status      tag.Key
+	pipeline    tag.Key
+	pipelineRun tag.Key
+	pod         tag.Key
+}
+
+// NewRecorder creates a new metrics recorder instance
+// to log the TaskRun related metrics
+func NewRecorder() (*Recorder, error) {
+	r := &Recorder{
+		initialized: true,
+	}
+
+	task, err := tag.NewKey("task")
+	if err != nil {
+		return nil, err
+	}
+	r.task = task
+
+	taskRun, err := tag.NewKey("taskrun")
+	if err != nil {
+		return nil, err
+	}
+	r.taskRun = taskRun
+
+	namespace, err := tag.NewKey("namespace")
+	if err != nil {
+		return nil, err
+	}
+	r.namespace = namespace
+
+	status, err := tag.NewKey("status")
+	if err != nil {
+		return nil, err
+	}
+	r.status = status
+
+	pipeline, err := tag.NewKey("pipeline")
+	if err != nil {
+		return nil, err
+	}
+	r.pipeline = pipeline
+
+	pipelineRun, err := tag.NewKey("pipelinerun")
+	if err != nil {
+		return nil, err
+	}
+	r.pipelineRun = pipelineRun
+
+	pod, err := tag.NewKey("pod")
+	if err != nil {
+		return nil, err
+	}
+	r.pod = pod
+
+	err = view.Register(
+		&view.View{
+			Description: trDuration.Description(),
+			Measure:     trDuration,
+			Aggregation: trDistribution,
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.status},
+		},
+		&view.View{
+			Description: prTRDuration.Description(),
+			Measure:     prTRDuration,
+			Aggregation: prTRLatencyDistribution,
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.status, r.pipeline, r.pipelineRun},
+		},
+		&view.View{
+			Description: trCount.Description(),
+			Measure:     trCount,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{r.status},
+		},
+		&view.View{
+			Description: runningTRsCount.Description(),
+			Measure:     runningTRsCount,
+			Aggregation: view.LastValue(),
+		},
+		&view.View{
+			Description: podLatency.Description(),
+			Measure:     podLatency,
+			Aggregation: view.LastValue(),
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.pod},
+		},
+	)
+
+	if err != nil {
+		r.initialized = false
+		return r, err
+	}
+
+	return r, nil
+}
+
+// DurationAndCount logs the duration of TaskRun execution and
+// count for number of TaskRuns succeed or failed
+// returns an error if its failed to log the metrics
+func (r *Recorder) DurationAndCount(tr *v1alpha1.TaskRun) error {
+	if !r.initialized {
+		return fmt.Errorf("ignoring the metrics recording for %s , failed to initialize the metrics recorder", tr.Name)
+	}
+
+	duration := time.Since(tr.Status.StartTime.Time)
+	if tr.Status.CompletionTime != nil {
+		duration = tr.Status.CompletionTime.Sub(tr.Status.StartTime.Time)
+	}
+
+	taskName := "anonymous"
```

Comment:
Can you file an issue to track a better solution for this?

In your previous comment you said the exporter drops labels with empty-string values, does that mean no data will be reported at all? Or just that it will be reported without a `taskName`? Because the latter is probably what we want, assuming we can query for data points that don't report a `taskName`.